### PR TITLE
Add new Application Callback (non-breaking) for early intercept of inbou...

### PR DESCRIPTION
**Pull request by martinadams**
...nd messages.

Add an IApplicationExt interface that extends IApplication and facilitates early
intercept of inbound messages on the part of the Application.

This is a bit similar to the late outbound intercept afforded by ToApp/ToAdmin,
and permits alteration of field values prior to their being validated by the engine.
The difference is that, with the new interface, only one callback is provided for
both App and Admin messages.

Usage of the existing IApplication interface is unaffected.

### CODE REVIEW

File: QuickFIXn/IApplicationExt.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+
+namespace QuickFix
+{
+    /// <summary>
+    /// This is the optional extension interface for processing session messages, and facilitates early interception of inbound messages.
+	/// 'Early', in this context, means after structure, length and checksum have been validated, but before any further validation has been performed.
```
Comment (gbirchmeier) [inline QuickFIXn/IApplicationExt.cs]: please fix this indent

Comment (gbirchmeier) [inline QuickFIXn/IApplicationExt.cs]: Would you mind fixing these indents to use not-tabs?  Thanks.

File: UnitTests/SessionTest.cs [hunk 2]
```diff
@@ -112,8 +112,49 @@ public void OnLogout(QuickFix.SessionID sessionID)
...
         {
         }
+        #endregion
+    }
+
+    class MockApplicationExt : QuickFix.IApplicationExt
+    {
+        public HashSet<string> InterceptedMessageTypes = new HashSet<string>();
+
+       #region Application Members
```
Comment (gbirchmeier) [inline UnitTests/SessionTest.cs]: indent is off by one

File: UnitTests/SessionTest.cs [hunk 3]
```diff
@@ -719,5 +761,35 @@ public void TestToAppResendDoNotSend()
...
+        {
+            var mockApp = new MockApplicationExt();
+            session = new QuickFix.Session(mockApp, new QuickFix.MemoryStoreFactory(), sessionID,
+             new QuickFix.DataDictionaryProvider(), new QuickFix.SessionSchedule(config), 0, new QuickFix.ScreenLogFactory(settings), new QuickFix.DefaultMessageFactory(), "blah");
+            session.SetResponder(responder);
+            session.CheckLatency = false;
+
+            Logon();
+            QuickFix.FIX42.NewOrderSingle order = new QuickFix.FIX42.NewOrderSingle(
+                 new QuickFix.Fields.ClOrdID("1"),
```
Comment (gbirchmeier) [inline UnitTests/SessionTest.cs]: please use 4-char indent

File: UnitTests/SessionTest.cs [hunk 4]
```diff
@@ -719,5 +761,35 @@ public void TestToAppResendDoNotSend()
...
             Assert.False(SENT_NOS());
         }
+
+
+        [Test]
+        public void TestApplicationExtension()
+        {
+            var mockApp = new MockApplicationExt();
+            session = new QuickFix.Session(mockApp, new QuickFix.MemoryStoreFactory(), sessionID,
+             new QuickFix.DataDictionaryProvider(), new QuickFix.SessionSchedule(config), 0, new QuickFix.ScreenLogFactory(settings), new QuickFix.DefaultMessageFactory(), "blah");
```
Comment (gbirchmeier) [inline UnitTests/SessionTest.cs]: please use 4 char indent

**Comment by gbirchmeier** [comment]
Looks good except for the little dumb indent things.

I do fear this feature a little bit, as I can imagine many misguided things that people will try to use it for.  

Can you add some doc comments at the top of `IApplicationExt.cs` to explain the IApplicationExt class and how to use it?  I don't think I want to officially document it in our tutorial (because the use of this should probably not be encouraged), but I think we should capture that so future maintainers and code-spelunkers will know what this is.

**Comment by martinadams** [comment]
Changes made as requested..

**Pull request merged**